docs/ql-design-patterns.md

```diff
@@ -0,0 +1,82 @@
+```
+
+use
+
+```ql
+/**
+ * <QLDoc...>
+ *
+ * Extend this class to refine existing API models. If you want to model new APIs,
+ * extend `MySpecialExpr::Range` instead.
+ */
+class MySpecialExpr extends Expr {
+  MySpecialExpr::Range range;
+
+  MySpecialExpr() { this = range }
+
+  /** <QLDoc...> */
+  int memberPredicate() { result = range.memberPredicate() }
+}
+
+/** Provides a class for modeling new <...> APIs. */
+module MySpecialExpr {
+  /**
+   * <QLDoc...>
+   *
+   * Extend this class to model new APIs. If you want to refine existing API models,
+   * extend `MySpecialExpr` instead.
+   */
+  abstract class Range extends Expr {
+    /** <QLDoc...> */
+    abstract int memberPredicate();
+  }
+}
+```
+with
+```ql
+class ConcreteSubclass extends MySpecialExpr::Range { ... }
+```
+
+</details>
+
+Note, previously we used to write `MySpecialExpr() { this = self }`, but we now recommend using `MySpecialExpr() { this = range }` instead to avoid anyone mistakenly thinking that `self` and `this` are synonyms in general.
+
+### Rationale
+
+Let's use an example from the Go libraries: https://github.com/github/codeql-go/blob/2ba9bbfd8ba1818b5ee9f6009c86a605189c9ef3/ql/src/semmle/go/Concepts.qll#L119-L157
+
+`EscapeFunction`, as the name suggests, models various APIs that escape meta-characters. It has a member-predicate `kind()` that tells you what sort of escaping the modelled function does. For example, if the result of that predicate is `"js"`, then this means that the escaping function is meant to make things safe to embed inside JavaScript.
+`EscapeFunction::Range` is subclassed to model various APIs, and `kind()` is implemented accordingly.
+But we can also subclass `EscapeFunction` to, as in the above example, talk about all JS-escaping functions.
+
+You can, of course, do the same without the `::Range` pattern, but it's a little cumbersome:
+If you only had an `abstract class EscapeFunction { ... }`, then `JsEscapeFunction` would need to be implemented in a slightly tricky way to prevent it from extending `EscapeFunction` (instead of refining it). You would have to give it a charpred `this instanceof EscapeFunction`, which looks useless but isn't. And additionally, you'd have to provide trivial `none()` overrides of all the abstract predicates defined in `EscapeFunction`. This is all pretty awkward, and we can avoid it by distinguishing between `EscapeFunction` and `EscapeFunction::Range`.
+
+
+## Importing all subclasses of abstract base class
+
+When providing an abstract class, you should ensure that all subclasses are included when the abstract class is (unless you have good reason not to). Otherwise you risk having different meanings of the abstract class depending on what you happen to import.
+
+One example where this _does not_ apply: `DataFlow::Configuration` and its variants are abstract, but we generally do not want to import all configurations into the same scope at once.
```

Comment:
```suggestion
## Abstract classes as open or closed unions

A class declared as `abstract` in QL represents a union of its direct subtypes (restricted by the intersections of its supertypes and subject to its characteristic predicate). Depending on context, we may want this union to be considered "open" or "closed".

An open union is generally used for extensibility. For example, the abstract classes suggested by the `::Range` design pattern are explicitly intended as extension hooks. As another example, the `DataFlow::Configuration` design pattern provides an abstract class that is intended to be subclassed as a configuration mechanism.

A closed union is a class for which we do not expect users of the library to add more values. Historically, we have occasionally modelled this as `abstract` classes in QL, but these days that would be considered an anti-pattern: Abstract classes that are intended to be closed behave in surprising ways when subclassed by library users, and importing libraries that include derived classes can invalidate compilation caches and subvert the meaning of the program.

Instead, in modern QL we prefer to model closed unions as dbscheme unions, type union aliases (for unions of `newtype` branches), or via explicit `this instanceof Type1 or this instanceof Type 2 or ...` in the characteristic predicate of the closed-union class. Alternatively, we may represent a closed union as an `abstract` class similarly to the `::Range` pattern, but making the `::Range` type private to our library.
```

Reply:
I might be misunderstanding, but it seems like we're addressing orthogonal problems.

I'm trying to highlight that when you are defining an open union such as an abstract class following the `::Range` pattern, you should ensure that all subclasses are imported (unless you have good reason not to). As a motivating example, when defining our `RemoteFlowSource::Range` in the Python rewrite, we _could_ have left it up to the queries to import the relevant ql modules that provide concrete subclasses of `RemoteFlowSource::Range` -- but that is an anti-pattern, and we should ensure that all concrete subclasses are imported by default ([which we do](
https://github.com/github/codeql/blob/76fd710946e5aa336721551517b0cff8daa21aff/python/ql/src/semmle/python/dataflow/new/RemoteFlowSources.qll#L4)). 

I don't quite understand the part about the closed unions, and that seems to be a different point deserving it's own heading. Can you maybe give an example of how the anti-pattern would look like? and what a "preferred" solution would be?

Reply:
I'm uncomfortable with this section as it stands, because it doesn't seem like good generic advice. It says "you should import all subtypes of an abstract type, unless you have a reason not to" -- but that's somewhat tautological. And even if we were to view it as restricted to just the `::Range` pattern, the very point of `::Range` is that end users can create new subtypes to extend the range *without modifying the core library*, and this by definition creates a situation where some subtypes are not imported by default.

My suggestion attempts to provide generic advice on when to use an abstract class (pretty much only if you are modelling an "open union", where you are anticipating that some clients want to contribute to the union). You're right that an example might help; does the following make it more understandable?

-------
As an example, suppose we want to define a `BinaryExpr` class, which has subtypes of `PlusExpr`, `MinusExpr`, and so on. Morally, this represents a closed union: We do not anticipate new kinds of `BinaryExpr` being added. Therefore, it would be undesirable to model it as an abstract class:

```ql
/** ANTI-PATTERN */
abstract class BinaryExpr extends Expr {
  Expr getLhs() { result = this.getChild(0) }
  Expr getRight() { result = this.getChild(1) }
}

class PlusExpr extends BinaryExpr {}
class MinusExpr extends BinaryExpr {}
...
```

Instead, the `BinaryExpr` class should be non-`abstract`, and we have the following options for specifying its extent:

- Define a dbscheme type `@binary_expr = @plus_expr | @minus_expr | ...` and add it as an additional super-class for `BinaryExpr`.
- Define a type alias `class RawBinaryExpr = @plus_expr | @minus_expr | ...` and add it as an additional super-class for `BinaryExpr`.
- Add a characteristic predicate of `BinaryExpr() { this instanceof PlusExpr or this instanceof MinusExpr or ... }`.

Intuitively, in a closed union, the union type is aware of all of its subtypes -- this is not the case for open unions, as in the `::Range` pattern.

------------

Does that make sense?

I think the lesson you are trying to convey perhaps deserves its own heading, and a more general phrasing. It is not specific to abstract classes. The guidance should be something like: "Importing new files can modify the behaviour of the standard library, by introducing new subtypes of `abstract` classes, by introducing new multiple inheritance relationships, or by overriding predicates. This can change query results and force evaluator cache misses."